### PR TITLE
feat: Add StoreView Kafka event consumer

### PR DESCRIPTION
**Pull request by bretthoerner**
First pass at consumer that will process StoreView messages: https://github.com/getsentry/sentry/blob/3fbfd80baf5ed245ff9d03754bf12ffaf912bf1d/src/sentry/web/api.py#L566-L585

Note that this is still part of a larger refactor and testing. This will not be run against production data at this point in time -- or if we do it will only be projects/orgs we opt-in.

### CODE REVIEW

File: src/sentry/runner/commands/storeconsumer.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+from __future__ import absolute_import, print_function
+
+import click
+import signal
+from django.conf import settings
+
+from sentry.runner.decorators import configuration
+
+
+@click.command()
```
Comment (bretthoerner) [inline src/sentry/runner/commands/storeconsumer.py]: The following (and the batching consumer) comes from the tried-and-true Snuba consumer, fwiw. Not a lot going on here.

File: src/sentry/event_consumer.py [hunk 2]
```diff
@@ -38,3 +47,20 @@ def process_event_from_kafka(message):
...
+        value = json.loads(message.value())
+        if value.get('should_process', False):
+            return value
+
+    def flush_batch(self, batch):
+        for event in batch:
+            try:
+                process_event_from_kafka(event)
+            except Exception:
+                logger.exception('Error processing event.')
```
Comment (bretthoerner) [inline src/sentry/event_consumer.py]: What goes in `process_message` vs `flush_batch` is a little hard to decide at this point in time. See the batching consumer code for an understanding:

https://github.com/getsentry/batching-kafka-consumer/blob/2596d2216b889197392aa9bec551f01d473adcc4/batching_kafka_consumer/__init__.py#L19-L82

So the thing you should notice right away is that none of our store code is (currently?) setup to operate on batches. *But* that doesn't mean we want to commit Kafka offsets for every message we process, nor do we want to use the dreaded auto-commit (aka "shrug") IMO.

So what I have here currently at least deserializes the data in `process_message`, and then when a batch-threshold is reached we actually iterate the results, writing (serially) to TSDB/Redis/Celery/whatever-else-StoreView-currently-touches.

This does mean the worker will operate in spurts, but I'm not sure it matters?

We could move `process_event_from_kafka` to `process_message`, and only use `flush_batch` to handle the Kafka commit. I think either way makes about as much sense with the code we have today.

Comment (bretthoerner) [reply src/sentry/event_consumer.py]: Shower thought: maybe we don't want to commit to Kafka every 500-1000ms, and I doubt we want to batch up unprocessed events for longer than that (creating even worse "spurts" of work).

So maybe I _should_ move the actual processing code into `process_message`, returning nothing but `True` and lean on `flush_batch` _solely_ for committing to Kafka every ~5 (10, 30, 60?) seconds. In the worst case a crash consumer means event are reprocessed, but we do de-dupe checking in this code path. And to repeat again: baby steps, in the future maybe we can refactor this code to actually operate on batches better, but that's not a task for today IMO.

**Comment by bretthoerner** [comment]
@getsentry/search Ready for a look if anyone has time.

**Review by tkaemming — CHANGES_REQUESTED** [review]
I don't really have any specific comments or questions on the implementation of this change itself. I do have questions around the change to batch processing.

1. `process_event_from_kafka` sets the [`event_manager._normalized` attribute to `True`](https://github.com/getsentry/sentry/blob/a077672d46f539cd00d06530ce32c5ca68c8939e/src/sentry/event_consumer.py#L36). I assume this is because the data is assumed to be written to the Kafka topic pre-normalized?
2. The contents of the [`process_event` method up to the `cache.set` call](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L95-L233) may be called any number of times if we fail to commit an offset. The salient (at least to my reading and interpretation, at least) side effects of this section are:
    1. Sending signals, including:
        1. [`event_received`](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L96), which has no receivers.
        2. [`event_filtered`](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L127), which is used to record [project](https://github.com/getsentry/getsentry/blob/6e9a3990eee37724b59dc087600aad05e74fe311/getsentry/receivers/usage.py#L145-L149) and [organization](https://github.com/getsentry/getsentry/blob/6e9a3990eee37724b59dc087600aad05e74fe311/getsentry/receivers/usage.py#L175-L179) usage data.
        3. [`event_dropped`](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L165), which is used to record [project](https://github.com/getsentry/getsentry/blob/6e9a3990eee37724b59dc087600aad05e74fe311/getsentry/receivers/usage.py#L139-L143) and [organization](https://github.com/getsentry/getsentry/blob/6e9a3990eee37724b59dc087600aad05e74fe311/getsentry/receivers/usage.py#L163-L167) usage data, as well for [spike projection](https://github.com/getsentry/getsentry/blob/6e9a3990eee37724b59dc087600aad05e74fe311/getsentry/receivers/usage.py#L169-L173).
    2. [Checking and updating quotas/rate limits](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L135-L139).
        * Is the fact that we're potentially checking this at a different rate than we were previously (How is this topic partitioned? Can an organization's events be split over multiple partitions consuming at different rates?) and batching up messages (shifting and compressing the time range where we are actually performing rate limiting checks) cause any messages to be discarded as a result?
    3. Incrementing TSDB counters, including `{project,organization,key}_total_{received,rejected}`, which are used mostly for organization and project API statistics endpoints, as well as [some internal usages](https://github.com/getsentry/getsentry/search?q=organization_total_received&unscoped_q=organization_total_received) (some of which might be related to billing, but I'm not really familiar enough to say offhand?)
        * Should the [`start_time`](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L98-L99) here reflect the Kafka message insertion time, not the current processing worker time? The difference between "occurred time" and "received time" already confuses people. Adding a third timeline seems like it's only going to make things worse? Is this going to have side effects on things like smart limits?
    4. Calling [`insert_data_to_database`](https://github.com/getsentry/sentry/blob/a077672d46f539cd00d06530ce32c5ca68c8939e/src/sentry/coreapi.py#L161-L183) which inserts the event into the cache and enqueues a Celery task. (This seems mostly straightfoward.)
3. As far as exception handling goes, `process_event_from_kafka` can throw  `APIUnauthorized` for a [variety of authentication related reasons.](https://github.com/getsentry/sentry/blob/a077672d46f539cd00d06530ce32c5ca68c8939e/src/sentry/coreapi.py#L121-L143) Exceptions can also be thrown via `process_event`, including `APIForbidden` for [inbound filters](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L132) or [duplicates](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L197), [`APIRateLimited`](https://github.com/getsentry/sentry/blob/fa27b007aaf779d72ae0f24e8ad274d654a70ac5/src/sentry/web/api.py#L172), or any other exception for known/expected reasons (network failures to TSDB, database issues retrieving project options, cache failures, failure to enqueue Celery task, for example) or unexpected reasons (bugs, etc.) All of these exception types are all treated the same and result in the message being skipped/discarded. Is this really desirable/intended?

**Comment by bretthoerner** [comment]
> I assume this is because the data is assumed to be written to the Kafka topic pre-normalized

Correct,

https://github.com/getsentry/sentry/blob/f5c696aa64ed8556c7a0117380b381172d1552e7/src/sentry/web/api.py#L550

The Relay will verify auth + normalize + publish (in the future). Normalize in Python even has a code path that re-uses the same Rust code, fwiw:

https://github.com/getsentry/sentry/blob/f5c696aa64ed8556c7a0117380b381172d1552e7/src/sentry/event_manager.py#L426

> The contents of the process_event method up to the cache.set call may be called any number of times if we fail to commit an offset

Correct. We could have an option in the batching consumer to be at-most-once, or we could add relocate the cache.get/set to happen sooner (obviously each of those options will mean dropping data on the floor in some scenarios).

> Sending signals [may be called any number of times if offsets are not written]

Correct.

> Checking and updating quotas/rate limits. Is the fact that we're potentially checking this at a different rate than we were previously... cause any messages to be discarded as a result?

Good point. As things are written: probably. It looks like we set `received` in normalize (aside: this is kind of weird to think about in a world of Relay, where the first Relay might normalize the event an hour before it makes it to our server or something...). And it looks like most rate limiting / TSDB counters are using a fresh `time()` during processing. I wonder if they should use `received` to account for Kafka backlogs/delays? Have to think about this more.

> Incrementing TSDB counters... Should the start_time here reflect the Kafka message insertion time, not the current processing worker time?

Heh, definitely related to my `received` question above. Using the Kafka insert time as yet-another-time is interesting (and I agree could be even more confusing).

> Calling insert_data_to_database [may be called any number of times if offsets are not written]

Correct.

> process_event_from_kafka can throw APIUnauthorized for a variety of authentication related reasons

I believe it won't in practice. Definitely a good call out to cleanup/clarify. In order to get this far (be published to Kafka) the auth has already been done in StoreView.dispatch (or Relay). The `project_key_from_auth` call is being used to get from the JSON blob back to the Key object we need during the process call.

> Exceptions can also be thrown via process_event, including APIForbidden for inbound filters or duplicates, APIRateLimited

Yeah, by publishing to Kafka we are definitely going into a "SDKs blindly get HTTP 200s for the most part" and processing can fail/reject/rate-limit asynchronously. According to Armin this is OK, as SDKs are (or should) be hiding failures from users today. So the future work would be to turn any rate-limit related failures into some kind of shared state (Redis?) that the StoreView/Relay can check *before* publishing to Kafka, so we at least return Retry-After where applicable.

As for events dropped because of duplication/filtering we have ideas for simple topics to record those stats (probably in ClickHouse as very-very-very reduced event blobs?) -- but that's far future work.

>  All of these exception types are all treated the same and result in the message being skipped/discarded. Is this really desirable/intended?

Honest question: does this differ from production today?

**Comment by bretthoerner** [comment]
As an aside that will effect a number of those comments: I *do* think I should move the `process_event_from_kafka` call into the `EventConsumerWorker.process_message` method, so that we process events as soon as we can `poll()` them. This should get us much closer to what we have today, keeping these bits in mind:

1. A graceful shutdown will still flush offsets.
2. An ungraceful exit will re-process N events.
3. I assume we want to partition by project. Not to do anything special today but to unlock/ensure intelligent project-based things we can do in the future. This will, however, mean some imbalance in partition message-rates.
4. Because 🐍 we should probably continue to overprovision partitions and consumers to (very roughly) "ensure" we are always processing near-realtime.

That said, all your questions/issues above still stand.

**Comment by tkaemming** [comment]
Thanks for the thoughtful and detailed reply.

I should have looked closer at the description and I would have been able to answer my own questions about the normalization happening prior to publishing to Kafka. There might be a better way to articulate this in code or comments, but I haven't put much thought into it.

> > The contents of the process_event method up to the cache.set call may be called any number of times if we fail to commit an offset
> 
> Correct. We could have an option in the batching consumer to be at-most-once, or we could add relocate the cache.get/set to happen sooner (obviously each of those options will mean dropping data on the floor in some scenarios).
> 
> edit: We could also check another cache key at the very start of the method that is composed of the topic, partition and offset -- as a "Someone for sure has attempted to process _this_ message" check. It would mean we would drop up to 1 event per non graceful exit, as the duplicate-message key will be set but the message probably won't make it through the entire processing pipeline.

For the sake of simplicity, I think it might be worth moving up the `cache.set` call, so that if processing fails for any reason, we don't have the potential to overcount for usage or rate limit data, especially if the data is going to possibly dropped in failure scenarios anyway without the potential for retry/replay. (This itself might have side effects that I haven't put much thought into, so don't take this as fully 
vetted feedback.) I'm not sure what the advantages would be to having a message marker for the Kafka log position specifically, compared to the event ID? They both seem like they'd solve the same problem, with the benefit of the existing `cache.set` approach also soft validating the `(project_id, event_id)` unique constraint.

> > Checking and updating quotas/rate limits. Is the fact that we're potentially checking this at a different rate than we were previously... cause any messages to be discarded as a result?
> 
> Good point. As things are written: probably. It looks like we set `received` in normalize (aside: this is kind of weird to think about in a world of Relay, where the first Relay might normalize the event an hour before it makes it to our server or something...). And it looks like most rate limiting / TSDB counters are using a fresh `time()` during processing. I wonder if they should use `received` to account for Kafka backlogs/delays? Have to think about this more.
> 
> > Incrementing TSDB counters... Should the start_time here reflect the Kafka message insertion time, not the current processing worker time?
> 
> Heh, definitely related to my `received` question above. Using the Kafka insert time as yet-another-time is interesting (and I agree could be even more confusing).

Yeah, I'm not sure about both of these either. I would similarly need to put more thought into it.

> > process_event_from_kafka can throw APIUnauthorized for a variety of authentication related reasons
> 
> I believe it won't in practice. Definitely a good call out to cleanup/clarify. In order to get this far (be published to Kafka) the auth has already been done in StoreView.dispatch (or Relay). The `project_key_from_auth` call is being used to get from the JSON blob back to the Key object we need during the process call.

This makes sense. I think it would be good to explicitly catch these exceptions (more on this later.)

> > Exceptions can also be thrown via process_event, including APIForbidden for inbound filters or duplicates, APIRateLimited
> 
> Yeah, by publishing to Kafka we are definitely going into a "SDKs blindly get HTTP 200s for the most part" and processing can fail/reject/rate-limit asynchronously. According to Armin this is OK, as SDKs are (or should) be hiding failures from users today. So the future work would be to turn any rate-limit related failures into some kind of shared state (Redis?) that the StoreView/Relay can check before publishing to Kafka, so we at least return Retry-After where applicable.
>
> As for events dropped because of duplication/filtering we have ideas for simple topics to record those stats (probably in ClickHouse as very-very-very reduced event blobs?) -- but that's far future work.

Yeah, I kind of figured as much regarding the response that the endpoint can return to the SDK, and what the SDK will actually be able to do with it (I don't think there is a reasonable alternative.) I think having some sort of feedback channel like you mentioned would make me feel better about this. The TSDB metrics and signals sort of do this today, but it sort of feels like they're happening at the wrong level of abstraction here (inlined in the processing code, rather than being consolidated in some higher level exception handling.) I guess this isn't a new development, it just seems different looking at it from a fresh perspective. Reworking that seems like it's probably out of the scope of this change, but I think worth considering at some point (if others have the same perspective, at least.)

> > All of these exception types are all treated the same and result in the message being skipped/discarded. Is this really desirable/intended?
> 
> Honest question: does this differ from production today?

:dread:

Good point. I think this feels bad because it *is* bad, but it's fair to say it's not a regression. Similarly, that's probably a justification for moving the cache set towards the beginning of processing, since it will at least retain similar behavior to the existing behavior (specifically, no retries.)

The things I'd still like to see here (mostly research and words, not code changes):

- Identify impact on billing and spike protection (via signals) due to time shifting (and potential for replay if the cache set stay as as-is.)
- Identify impact on rate limiting due to time shifting (and potential for replay if the cache set stays as as-is.
- Split error handling for expected (rate limiting, inbound filters, etc — generically the `API*` exceptions) and unexpected errors for clarity, even if they do the same thing right now. I think in the future we'd want to reconsider how error handling happens here (consolidating/cataloging all of the different failure modes/exception handlers at a higher level, retrying/republishing transient faiures, etc.)

**Comment by bretthoerner** [comment]
> I'm not sure what the advantages would be to having a message marker for the Kafka log position specifically, compared to the event ID? They both seem like they'd solve the same problem

It may be very minor, but the current code still increments things like "total received" if it's a dupe. So there is a very tiny difference between "is this a dupe the user sent" and "are we reprocessing the message because a consumer died." Probably not worth the confusion though.

Agree that moving `cache.set` up is probably the best move.

> The things I'd still like to see here (mostly research and words, not code changes):
> 
> * Identify impact on billing and spike protection (via signals) due to time shifting (and potential for replay if the cache set stay as as-is.)
> * Identify impact on rate limiting due to time shifting (and potential for replay if the cache set stays as as-is.
> * Split error handling for expected (rate limiting, inbound filters, etc — generically the API* exceptions) and unexpected errors for clarity, even if they do the same thing right now. I think in the future we'd want to reconsider how error handling happens here (consolidating/cataloging all of the different failure modes/exception handlers at a higher level, retrying/republishing transient faiures, etc.)

+1

**Comment by tkaemming** [comment]
Also, thinking about it more, I think there might be value in separating the concerns of "prevent replaying a specific Kafka message" and "prevent storing an event multiple times." I think you could make a reasonable argument that if somebody's custom SDK is running amok and repeatedly sending the same event ID, that should go against their quota (and be billed for it.) If we have an internal error that results in some messages being replayed, I don't think it's fair to penalize the user for that, and those seem like two separate problems.

If we're concerned about the cost of maintaining that extra state (a different nonce for topic/partition/offset), we can move the existing cache set nonce earlier in the process like we had talked about (without changing the way the key is generated), and that'll have the same effect in most cases, but will not penalize malfunctioning/misbehaving clients adding undue pressure on the ingestion systems.

File: src/sentry/event_consumer.py [hunk 3]
```diff
@@ -38,3 +46,31 @@ def process_event_from_kafka(message):
...
+
+        # TODO: The BatchingKafkaConsumer API only flushes batches (and Kafka offsets)
+        # if there are non-None values returned from `process_message` that should be
+        # flushed. However, our existing code doesn't operate on batches, so we eagerly
+        # process them as soon as they are received from Kafka, and return True so that
+        # there is something in the batch for the BatchingKafkaConsumer to flush (along
+        # with offsets). We should maybe improve the BatchingKafkaConsumer API to allow
+        # for flushing consumed offsets without having to "fake" a batch.
```
Comment (bretthoerner) [inline src/sentry/event_consumer.py]: https://github.com/getsentry/batching-kafka-consumer/pull/1

**Comment by bretthoerner** [comment]
> I'm going to take a stab at 3 (maybe in a PR ontop of this one if I get anywhere)

This actually seemed pretty small... currently in this PR at: https://github.com/getsentry/sentry/pull/11061/commits/8013c3b11ce084eb60404baac3c75793d6f7a200

**Pull request closed without merging**